### PR TITLE
docs: Properly alphabetize output of CLI --help option (backport 12331)

### PR DESCRIPTION
**Pull request by Minh20**
Alphebetize CLI --help output #3614

### CODE REVIEW

**Comment by PastaPastaPasta** [comment]
haven't taken a look yet, however, in the future, try to avoid closing a PR just to open a new one, because now the discussion we had on the old PR we can't view here. Also it kindof just clutters up the repo. Granted, I did this exact same thing when I was getting started because git was confusing and I didn't understand it, so if you need help with something like git stuff, let us know and we can likely help you

**Comment by Minh20** [comment]
> haven't taken a look yet, however, in the future, try to avoid closing a PR just to open a new one, because now the discussion we had on the old PR we can't view here. Also it kindof just clutters up the repo. Granted, I did this exact same thing when I was getting started because git was confusing and I didn't understand it, so if you need help with something like git stuff, let us know and we can likely help you

Thanks for the feedback and offering help for the git stuff. I'll pay attention for future pull requests.

File: src/init.cpp [hunk 1]
```diff
@@ -543,72 +546,77 @@ std::string HelpMessage(HelpMessageMode mode)
...
         strUsage += HelpMessageOpt("-checkmempool=<n>", strprintf("Run checks every <n> transactions (default: %u)", defaultChainParams->DefaultConsistencyChecks()));
         strUsage += HelpMessageOpt("-checkpoints", strprintf("Disable expensive verification for known chain history (default: %u)", DEFAULT_CHECKPOINTS_ENABLED));
         strUsage += HelpMessageOpt("-deprecatedrpc=<method>", "Allows deprecated RPC method(s) to be used");
+        strUsage += HelpMessageOpt("-disablesafemode", strprintf("Disable safemode, override a real safe mode event (default: %u)", DEFAULT_DISABLE_SAFEMODE));
         strUsage += HelpMessageOpt("-dropmessagestest=<n>", "Randomly drop 1 of every <n> network messages");
-        strUsage += HelpMessageOpt("-stopafterblockimport", strprintf("Stop running after importing blocks from disk (default: %u)", DEFAULT_STOPAFTERBLOCKIMPORT));
-        strUsage += HelpMessageOpt("-stopatheight", strprintf("Stop running after reaching the given height in the main chain (default: %u)", DEFAULT_STOPATHEIGHT));
-
+        strUsage += HelpMessageOpt("-fuzzmessagestest=<n>", "Randomly fuzz 1 of every <n> network messages");
```
Comment (UdjinM6) [inline src/init.cpp:576]:
```suggestion
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Comment by PastaPastaPasta** [comment]
feature_shutdown.py failure here: https://gitlab.com/dashpay/dash/-/jobs/711537367 is weird... No need for you to worry about this @Minh20

File: src/init.cpp [hunk 2]
```diff
@@ -473,33 +473,35 @@ std::string HelpMessage(HelpMessageMode mode)
...
 #endif
     strUsage += HelpMessageOpt("-prune=<n>", strprintf(_("Reduce storage requirements by enabling pruning (deleting) of old blocks. This allows the pruneblockchain RPC to be called to delete specific blocks, and enables automatic pruning of old blocks if a target size in MiB is provided. This mode is incompatible with -txindex, -rescan and -disablegovernance=false. "
             "Warning: Reverting this setting requires re-downloading the entire blockchain. "
-            "(default: 0 = disable pruning blocks, 1 = allow manual pruning via RPC, >=%u = automatically prune block files to stay under the specified target size in MiB)"), MIN_DISK_SPACE_FOR_BLOCK_FILES / 1024 / 1024));
-    strUsage += HelpMessageOpt("-reindex-chainstate", _("Rebuild chain state from the currently indexed blocks"));
-    strUsage += HelpMessageOpt("-reindex", _("Rebuild chain state and block index from the blk*.dat files on disk"));
+            "(default: 0 = disable pruning blocks, 1 = allow manual pruning via RPC, >%u = automatically prune block files to stay under the specified target size in MiB)"), MIN_DISK_SPACE_FOR_BLOCK_FILES / 1024 / 1024));
```
Comment (UdjinM6) [inline src/init.cpp:490]:
```suggestion
            "(default: 0 = disable pruning blocks, 1 = allow manual pruning via RPC, >=%u = automatically prune block files to stay under the specified target size in MiB)"), MIN_DISK_SPACE_FOR_BLOCK_FILES / 1024 / 1024));
```

File: src/dash-cli.cpp [hunk 3]
```diff
@@ -41,18 +41,18 @@ std::string HelpMessageCli()
...
     strUsage += HelpMessageOpt("-rpcconnect=<ip>", strprintf(_("Send commands to node running on <ip> (default: %s)"), DEFAULT_RPCCONNECT));
+    strUsage += HelpMessageOpt("-rpcpassword=<pw>", _("Password for JSON-RPC connections"));
     strUsage += HelpMessageOpt("-rpcport=<port>", strprintf(_("Connect to JSON-RPC on <port> (default: %u or testnet: %u)"), defaultBaseParams->RPCPort(), testnetBaseParams->RPCPort()));
-    strUsage += HelpMessageOpt("-rpcwait", _("Wait for RPC server to start"));
     strUsage += HelpMessageOpt("-rpcuser=<user>", _("Username for JSON-RPC connections"));
-    strUsage += HelpMessageOpt("-rpcpassword=<pw>", _("Password for JSON-RPC connections"));
-    strUsage += HelpMessageOpt("-rpcclienttimeout=<n>", strprintf(_("Timeout in seconds during HTTP requests, or 0 for no timeout. (default: %d)"), DEFAULT_HTTP_CLIENT_TIMEOUT));
+    strUsage += HelpMessageOpt("-rpcwait", _("Wait for RPC server to start"));
+    strUsage += HelpMessageOpt("-rpcwallet=<walletname>", _("Send RPC for non-default wallet on RPC server (argument is wallet filename in dashd directory, required if dashd/-Qt runs with multiple wallets)"));
```
Comment (UdjinM6) [inline src/dash-cli.cpp:51]:
```suggestion
    strUsage += HelpMessageOpt("-rpcwallet=<walletname>", _("Send RPC for non-default wallet on RPC server (needs to exactly match corresponding -wallet option passed to dashd)"));
```

File: src/dash-cli.cpp [hunk 4]
```diff
@@ -41,18 +41,18 @@ std::string HelpMessageCli()
...
+    strUsage += HelpMessageOpt("-stdin", _("Read extra arguments from standard input, one per line until EOF/Ctrl-D (recommended for sensitive information such as passphrases)"));
```
Comment (UdjinM6) [inline src/dash-cli.cpp:52]:
```suggestion
    strUsage += HelpMessageOpt("-stdin", _("Read extra arguments from standard input, one per line until EOF/Ctrl-D (recommended for sensitive information such as passphrases).  When combined with -stdinrpcpass, the first line from standard input is used for the RPC password."));
```

File: src/init.cpp [hunk 5]
```diff
@@ -453,13 +453,13 @@ std::string HelpMessage(HelpMessageMode mode)
...
     strUsage += HelpMessageOpt("-blocknotify=<cmd>", _("Execute command when the best block changes (%s in cmd is replaced by block hash)"));
+    strUsage += HelpMessageOpt("-blockreconstructionextratxn=<n>", strprintf(_("Extra transactions to keep in memory for compact block reconstructions (default: %u)"), DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN));
     if (showDebug)
         strUsage += HelpMessageOpt("-blocksonly", strprintf(_("Whether to operate in a blocks only mode (default: %u)"), DEFAULT_BLOCKSONLY));
-    strUsage +=HelpMessageOpt("-assumevalid=<hex>", strprintf(_("If this block is in the chain assume that it and its ancestors are valid and potentially skip their script verification (0 to verify all, default: %s, testnet: %s)"), defaultChainParams->GetConsensus().defaultAssumeValid.GetHex(), testnetChainParams->GetConsensus().defaultAssumeValid.GetHex()));
-    strUsage += HelpMessageOpt("-conf=<file>", strprintf(_("Specify configuration file. Relative paths will be prefixed by datadir location. (default: %s)"), BITCOIN_CONF_FILENAME));
+    strUsage += HelpMessageOpt("-conf=<file>", strprintf(_("Specify configuration file (default: %s)"), BITCOIN_CONF_FILENAME));
```
Comment (UdjinM6) [inline src/init.cpp:462]:
```suggestion
    strUsage += HelpMessageOpt("-conf=<file>", strprintf(_("Specify configuration file. Relative paths will be prefixed by datadir location. (default: %s)"), BITCOIN_CONF_FILENAME));
```

File: src/init.cpp [hunk 6]
```diff
@@ -473,33 +473,35 @@ std::string HelpMessage(HelpMessageMode mode)
...
-    strUsage += HelpMessageOpt("-timestampindex", strprintf(_("Maintain a timestamp index for block hashes, used to query blocks hashes by a range of timestamps (default: %u)"), DEFAULT_TIMESTAMPINDEX));
+    strUsage += HelpMessageOpt("-reindex", _("Rebuild chain state and block index from the blk*.dat files on disk"));
+    strUsage += HelpMessageOpt("-reindex-chainstate", _("Rebuild chain state from the currently indexed blocks"));
     strUsage += HelpMessageOpt("-spentindex", strprintf(_("Maintain a full spent index, used to query the spending txid and input index for an outpoint (default: %u)"), DEFAULT_SPENTINDEX));
+    strUsage += HelpMessageOpt("-timestampindex", strprintf(_("Maintain a timestamp index for block hashes, used to query blocks hashes by a range of timestamps (default: %u)"), DEFAULT_TIMESTAMPINDEX));
+    strUsage += HelpMessageOpt("-txindex", strprintf(_("Maintain a full transaction index, used by the getrawtransaction rpc call (default: %u)"), DEFAULT_TXINDEX));
+
```
Comment (UdjinM6) [inline src/init.cpp:504]:
```suggestion
```

File: src/init.cpp [hunk 7]
```diff
@@ -543,72 +546,70 @@ std::string HelpMessage(HelpMessageMode mode)
...
+    strUsage += HelpMessageOpt("-zmqpubrawinstantsenddoublespend=<address>", _("Enable publish raw transactions of attempted InstantSend double spend in <address>"));
     strUsage += HelpMessageOpt("-zmqpubrawtx=<address>", _("Enable publish raw transaction in <address>"));
     strUsage += HelpMessageOpt("-zmqpubrawtxlock=<address>", _("Enable publish raw transaction (locked via InstantSend) in <address>"));
-    strUsage += HelpMessageOpt("-zmqpubrawinstantsenddoublespend=<address>", _("Enable publish raw transactions of attempted InstantSend double spend in <address>"));
 #endif
 
     strUsage += HelpMessageGroup(_("Debugging/Testing options:"));
-    strUsage += HelpMessageOpt("-uacomment=<cmt>", _("Append comment to the user agent string"));
     if (showDebug) {
+        strUsage += HelpMessageOpt("-checkblockindex", strprintf("Do a full consistency check for mapBlockIndex, setBlockIndexCandidates, chainActive and mapBlocksUnlinked occasionally. Also sets -checkmempool (default: %u)", defaultChainParams->DefaultConsistencyChecks()));
```
Comment (UdjinM6) [inline src/init.cpp:568]:
```suggestion
        strUsage += HelpMessageOpt("-checkblockindex", strprintf("Do a full consistency check for mapBlockIndex, setBlockIndexCandidates, chainActive and mapBlocksUnlinked occasionally. (default: %u)", defaultChainParams->DefaultConsistencyChecks()));
```

File: src/init.cpp [hunk 8]
```diff
@@ -638,20 +639,20 @@ std::string HelpMessage(HelpMessageMode mode)
...
-    strUsage += HelpMessageOpt("-rpcbind=<addr>[:port]", _("Bind to given address to listen for JSON-RPC connections. Do not expose the RPC server to untrusted networks such as the public internet! This option is ignored unless -rpcallowip is also passed. Port is optional and overrides -rpcport. Use [host]:port notation for IPv6. This option can be specified multiple times (default: 127.0.0.1 and ::1 i.e., localhost, or if -rpcallowip has been specified, 0.0.0.0 and :: i.e., all addresses)"));
+    strUsage += HelpMessageOpt("-rpcallowip=<ip>", _("Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0) or a network/CIDR (e.g. 1.2.3.4/24). This option can be specified multiple times"));
+    strUsage += HelpMessageOpt("-rpcauth=<userpw>", _("Username and hashed password for JSON-RPC connections. The field <userpw> comes in the format: <USERNAME>:<SALT>$<HASH>. A canonical python script is included in share/rpcuser. The client then connects normally using the rpcuser=<USERNAME>/rpcpassword=<PASSWORD> pair of arguments. This option can be specified multiple times"));
```
Comment (UdjinM6) [inline src/init.cpp:644]:
```suggestion
    strUsage += HelpMessageOpt("-rpcauth=<userpw>", _("Username and hashed password for JSON-RPC connections. The field <userpw> comes in the format: <USERNAME>:<SALT>$<HASH>. A canonical python script is included in share/rpcauth. The client then connects normally using the rpcuser=<USERNAME>/rpcpassword=<PASSWORD> pair of arguments. This option can be specified multiple times"));
```

File: src/init.cpp [hunk 9]
```diff
@@ -638,20 +639,20 @@ std::string HelpMessage(HelpMessageMode mode)
...
+    strUsage += HelpMessageOpt("-rpcbind=<addr>[:port]", _("Bind to given address to listen for JSON-RPC connections. This option is ignored unless -rpcallowip is also passed. Port is optional and overrides -rpcport. Use [host]:port notation for IPv6. This option can be specified multiple times (default: 127.0.0.1 and ::1 i.e., localhost, or if -rpcallowip has been specified, 0.0.0.0 and :: i.e., all addresses)"));
```
Comment (UdjinM6) [inline src/init.cpp:645]:
```suggestion
    strUsage += HelpMessageOpt("-rpcbind=<addr>[:port]", _("Bind to given address to listen for JSON-RPC connections. Do not expose the RPC server to untrusted networks such as the public internet! This option is ignored unless -rpcallowip is also passed. Port is optional and overrides -rpcport. Use [host]:port notation for IPv6. This option can be specified multiple times (default: 127.0.0.1 and ::1 i.e., localhost, or if -rpcallowip has been specified, 0.0.0.0 and :: i.e., all addresses)"));
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Comment by UdjinM6** [comment]
Pls see https://github.com/UdjinM6/dash/commit/769f08ff51ef981a435bc2cdf1c08e360620f166 and maybe https://github.com/UdjinM6/dash/commit/bc2ac58418d103ff39b654507869ade26bb54db1 (or just sort options in `src/wallet/init.cpp` properly instead but imo wallet has so many options already that it kind of makes sense to split them in few more groups)

**Comment by Minh20** [comment]
> Pls see [UdjinM6@769f08f](https://github.com/UdjinM6/dash/commit/769f08ff51ef981a435bc2cdf1c08e360620f166) and maybe [UdjinM6@bc2ac58](https://github.com/UdjinM6/dash/commit/bc2ac58418d103ff39b654507869ade26bb54db1) (or just sort options in `src/wallet/init.cpp` properly instead but imo wallet has so many options already that it kind of makes sense to split them in few more groups)

**Comment by UdjinM6** [comment]
Still waiting for `src/wallet/init.cpp` updates...

**Comment by PastaPastaPasta** [comment]
Look https://github.com/dashpay/dash/pull/3681/files, you messed some stuff up

**Comment by PastaPastaPasta** [comment]
@Minh20 see commits https://github.com/UdjinM6/dash/commit/bc2ac58418d103ff39b654507869ade26bb54db1 and https://github.com/UdjinM6/dash/commit/769f08ff51ef981a435bc2cdf1c08e360620f166

**Review by UdjinM6 — APPROVED** [review]
ACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK, looks good

**Pull request merged**